### PR TITLE
Add precalculateValue() to AuxKernel

### LINKED ISSUE #7471: Add precalculateValue() to AuxKernel
### Description of the enhancement or error report
Some AuxKernels are operating strictly on a per-element basis, however `compute()` still calls `computeValue()` for all QPs and test functions. 

### Rationale for the enhancement or information for reproducing the error
Kernels have `precalculateResidual()` to perform once per element calculations, so I propose `precalculateValue()` for AuxKernels.

### Identified impact
Five AuxKernels in PhaseField could use this (and AuxKernels in Moose apps).

### PR DESCRIPTION
**Pull request by dschwen**
Just putting this out here...

Closes #7471

### CODE REVIEW

**Comment by permcody** [comment]
👍

**Comment by friedmud** [comment]
Wait - WTH is this?

Why did this go in without any discussion?

This changed a high profile and highly used API!  It should have been marked as `High Risk` and received scrutiny by the whole team....

**Comment by friedmud** [comment]
Why don't we just directly support this code path in `AuxKernel` instead?  i.e.`computeQpValue()` could only be called _once_ per element and you return one value... and no averaging is done.  It would be a lot more efficient than what's going on here.

Another option is that objects that want to do this just override `compute()` and call `_var.setNodalValue()` themselves.... no need to add to the API or do any extra logic or averaging or projecting or anything...

Or we could have `ElementConstantAuxKernel` that overrides `compute()` and calls `computeElemValue()` once to make a nice interface.

Basically... there should have been some discussion here...

**Comment by dschwen** [comment]
This was done to mirror `precalculateResidual()` in `Kernel`.

> Why don't we just directly support this code path in AuxKernel instead? i.e.computeQpValue() could only be called once per element and you return one value

AuxKernel does not know if the derived class returns the same value for each qp.

> Another option is that objects that want to do this just override compute() and call

Nah, then you have to cover all the cases of elemental and nodal variables yourself

> Or we could have ElementConstantAuxKernel that overrides compute() and calls computeElemValue() once to make a nice interface.

That seems like a good idea. We could restrict such kernels to constant monomial variables. We might as well keep the `computeValue()` name for the method, though.

**Comment by friedmud** [comment]
Ok - I opened #7552 .

I've had this idea before - but never captured it as a ticket.
